### DISCUSSION TITLE
Semantic Highlighting: extend beyond SemanticTokenTypes

### DISCUSSION THREAD

**Discussion by RobertWalterU3d** [eclipse-langium/langium#644] [Q&A] [answered]
Hello,

I try to understand semantic highlighting. I managed to add a `MyDSLSemanticTokenProvider` and change the highlighting based on the passed node's properties.
What I would like to do is have a different *formatting* for my language's cross references. I want them to be italic. So, this might be out-of-scope for what the semantic highlighting can even provide.
Still, when I look at the declaration of `SemanticTokenTypes` in the vscode-languageserver-protocol module, it hints at the option of adding additional token types:

![image](https://user-images.githubusercontent.com/63954267/186146018-9ce9bc8a-85c7-4a48-afce-88a5ae52b72e.png)

Let's forget for a moment that my goal is to make these italic, but let's assume I want to introduce the token type "reference". How would different themes handle this new token type, even? Would I have to define its highlighting as a default and they would use that as a fall back, or would they ignore this token type?

**Comment by msujew** [comment, marked as answer]
Hey @RobertWalterU3d,

as the LSP spec already says, it depends on the client capabilities. vscode (as one example of such a language client) would just ignore a custom `SemanticTokenTypes` entry by default. You can add custom type behavior through a theming extension though. See this [stackoverflow thread](https://stackoverflow.com/a/50714195).

> So, this might be out-of-scope for what the semantic highlighting can even provide.

You're right, Langium or the language server protocol in general aren't designed to provide this information. It's always up to the client how they want to display certain tokens. 

> What I would like to do is have a different _formatting_ for my language's cross references

On another note, _formatting_ in the language server context is usually reserved for code formatting. _Token styling_ might be the better term for what you're looking for.

**Reply by RobertWalterU3d** [reply]
Thanks Mark,
and you are right about the wording, I meant styling, but wrote formatting.

Best,
Robert
